[pytorch][nnc] mobile nnc backend skeleton

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56852 [pytorch][nnc] mobile nnc backend skeleton**
* #56851 [pytorch][nnc] protocol classes to persist the context for compiled functions



This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

ljk53 committed:
[pytorch][nnc] mobile nnc backend skeleton

This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"


This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"


This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"


This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"


This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/mobile/nnc/backend.cpp:
```diff
@@ -0,0 +1,61 @@
+};
+
+namespace {
+static const auto cls = torch::jit::backend<NNCBackend>("nnc");
```
What about the corresponding preprocess function?

ljk53 replied:
I think the new interface allows us to register the backend without preprocess method. And I think it's correct - the runtime binary itself doesn't need to have code to do AOT compilation.
The AOT part is in the stacked Phabricator diff: D27897136

kimishpatel replied:
> The AOT part is in the stacked Phabricator diff: D27897136

Thats what I was looking for. Thanks!

kimishpatel reviewed (commented):
Dont see the preprocess function?

kimishpatel reviewed (approved):
Looks good to me.

iseeyuan reviewed (approved):
LGTM

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] mobile nnc backend skeleton"




This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

[ghstack-poisoned]
